app/lib/features/bluetooth/backend/bluetooth_device.dart

```diff
@@ -0,0 +1,169 @@
+part of 'bluetooth_backend.dart';
+
+/// Wrapper class for bluetooth implementations to generically expose required functionality
+///
+/// This class can't be used directly, it should be implemented by a backend
+abstract class BluetoothDevice<BM extends BluetoothManager, BS extends BluetoothService, BC extends BluetoothCharacteristic, BackendDevice> with TypeLogger {
+  /// Create a new BluetoothLowEnergyDevice
+  ///
+  /// [_manager] Manager the device belongs to
+  /// [_source] Device implementation of the current backend
+  BluetoothDevice(this._manager, this._source) {
+    logger.finer('init device: $this');
+  }
+
+  /// BluetoothManager this device belongs to
+  late final BM _manager;
+
+  /// Corresponding BluetoothManager
+  BM get manager => _manager;
+
+  /// Original source device as returned by the backend
+  late final BackendDevice _source;
+
+  /// Original source device as returned by the backend
+  BackendDevice get source => _source;
+
+  /// (Unique?) id of the device
+  String get deviceId;
+
+  /// Name of the device
+  String get name;
+
+  /// Memoized service list for the device
+  List<BS>? _services;
+
+  bool _isConnected = false;
+  StreamSubscription<BluetoothConnectionState>? _connectionListener;
+
+  /// Whether the device is connected
+  bool get isConnected => _isConnected;
+
+  /// Stream to listen to for connection state changes after connecting to a device
+  Stream<BluetoothConnectionState> get connectionStream;
+
+  /// Backend implementation to connect to the device
+  Future<void> backendConnect();
+  /// Backend implementation to disconnect to the device
+  Future<void> backendDisconnect();
+
+  /// Array of disconnect callbacks
+  ///
+  /// Disconnect callbacks are processed in reverse order, i.e. the latest added callback is executed as first. Callbacks
+  /// can return true to indicate they have fully handled the disconnect. This will then also stop executing any remaining
+  /// callbacks.
+  @protected
+  final List<bool Function(bool wasConnected)> disconnectCallbacks = [];
+
+  /// Connect to the device
+  ///
+  /// Always call disconnect when ready after calling connect
+  Future<bool> connect({ VoidCallback? onConnect, bool Function(bool wasConnected)? onDisconnect, ValueSetter<Object>? onError, int maxTries = 5 }) {
+    final completer = Completer<bool>();
+    int connectTry = 1;
+
+    logger.finer('connect: Init');
+
+    if (onDisconnect != null) {
+      disconnectCallbacks.add(onDisconnect);
+    }
+
+    /// Local helper util to only complete the completer when it's not completed yet
```

Comment:
Since that code is trivial in-lining is fine. Someone reading that code has an easier time understanding what the if statement does than trying to decode the function name and read the doc or jump to its definition.